### ISSUE TITLE
InvalidOperationException in UmbracoMapper due to concurrent modification of non-thread-safe inner dictionary

### ISSUE THREAD

**Issue by albinj** [umbraco/Umbraco-CMS#15830]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

13.1.1

### Bug summary

Concurrent requests utilizing `UmbracoMapper.cs` result in an `InvalidOperationException` due to a race condition affecting a non-thread-safe inner Dictionary within the _ctors field.  

### Specifics

This concurrency issue was encountered when mapping one type of object to multiple other objects (multiple mapping definitions) and performing multiple concurrent requests when Umbraco boots to a url with a view that did all the mappings at once.

```
Error: Umbraco.Cms.Web.Common.PublishedModels.FrontPage => Site.MyModel
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
at Umbraco.Cms.Core.Mapping.UmbracoMapper.GetCtor(Type sourceType, Type targetType)
at Umbraco.Cms.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, Type sourceType, MapperContext context)
at Umbraco.Cms.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, MapperContext context)
at Umbraco.Cms.Core.Mapping.MapperContext.Map[TTarget](Object source)
```

The stack trace points to the `UmbracoMapper.GetCtor `method as the origin of the exception, indicating that the issue occurs during the retrieval of constructor mappings in concurrent scenarios.

### Steps to reproduce

1. Implement an `UmbracoMapper` mapping, for example, mapping from `Umbraco.Cms.Web.Common.PublishedModels.FrontPage` to `Site.MyModel`.
2. Use the mapper in a razor view
3. Simultaneously send multiple requests during Umbraco startup to the view rendering the mapped models using tools like curl to simulate concurrent access.
4. Monitor the application for any instances of `InvalidOperationException` indicating corrupted state due to concurrent modifications.

### Expected result / actual result

**Expected Result:**
`UmbracoMapper` should safely handle concurrent mapping requests, ensuring that all internal collections maintain thread safety and integrity, thus preventing any `InvalidOperationException` related to concurrent collection modifications.

**Actual Result:**
When subjected to concurrent requests, `UmbracoMapper` encounters an `InvalidOperationException` due to unsafe modifications on the non-thread-safe inner Dictionary within `_ctors`. This issue disrupts normal application operation, specifically during the mapping of objects in multi-threaded environments.

**Comment by albinj** [comment]
Changing the inner Dictionary with a ConcurrentDictionary seems to fix the issue.

**Comment by albinj** [comment]
PR https://github.com/umbraco/Umbraco-CMS/pull/15849

**Comment by kows** [comment]
Still got this in 13.2.2 

System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at Umbraco.Cms.Core.Mapping.UmbracoMapper.GetCtor(Type sourceType, Type targetType)
   at Umbraco.Cms.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, Type sourceType, MapperContext context)
   at Umbraco.Cms.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, MapperContext context)
   at Umbraco.Cms.Core.Mapping.UmbracoMapper.Map[TTarget](Object source)
   at AspNetCoreGeneratedDocument.Views_Master.ExecuteAsync() in D:\a\1\s\Libs\Client.Umbraco\Views\Master.cshtml:line 21

**Comment by albinj** [comment]
@kows Its been fixed in the upcoming release, 13.3.0.
